src/vmm/src/device_manager/acpi.rs

```diff
@@ -42,44 +57,90 @@ impl ACPIDeviceManager {
 
 impl Aml for ACPIDeviceManager {
     fn append_aml_bytes(&self, v: &mut Vec<u8>) {
-        // If we have a VMGenID device, create the AML for the device and GED interrupt handler
-        self.vmgenid.as_ref().inspect(|vmgenid| {
-            // AML for GED
-            aml::Device::new(
-                "_SB_.GED_".into(),
-                vec![
-                    &aml::Name::new("_HID".into(), &"ACPI0013"),
-                    &aml::Name::new(
-                        "_CRS".into(),
-                        &aml::ResourceTemplate::new(vec![&aml::Interrupt::new(
+        // Depending on what devices are available, generate AML for GED.
+
+        let locked_container = self.cpu_container.lock().expect("Poisoned lock");
+        match self.vmgenid.as_ref() {
```

Comment:
I agree, it is very similar. I tried to prevent this whilst I was working on it, but I'm yet to find a suitable solution. I keep running into issues with lifetimes. Looking into it.

Reply:
I'm not sure there is a way to do this. The way the AML is implemented, using `&dyn Aml`, we cannot meaningfully store any AML implementing structs. If you try to store one and later try and convert it to AML bytes later, you receive a `temporary value dropped while borrowed` error. Instead we must fully specify the device and convert it to AML bytes immediately.

This is what I've understood from looking into it, I've tried many different permutations of this and they all lead to the same error above. Obviously though, my Rust knowledge is limited, so if there is a way to do this that I'm missing, please let me know :)

Reply:
James is right here. Most of the `aml` module methods get as an argument a `&'a dyn Aml` arguments for the child objects. (I assume the authors did that for some sort of optimization(?)). As a result, you can use this in a sort of builder pattern which needs to end to a call to `append_aml_bytes`:

```Rust
aml::Device::new(.... vec[ &aml::SomeOtherObject(...), ...]).append_aml_bytes(v);l
```

which is unfortunate, because it makes it extremely hard to write nice/readable code.

Reply:
Ouch, alright :(